File: fastapi-apps/app_figma_component_inspector/app/utils/error_handler.py
"""
Error Handler Utilities

Custom exception handlers for FastAPI.
Matches error format from Fastify backend.
"""

from fastapi import Request, status
from fastapi.exceptions import RequestValidationError
from fastapi.responses import JSONResponse
from pydantic import ValidationError
from sqlalchemy.exc import SQLAlchemyError

from app.config import settings


async def validation_exception_handler(request: Request, exc: RequestValidationError):
    """
    Handle Pydantic validation errors.

    Returns 400 with validation details.
    """
    errors = []
    for error in exc.errors():
        errors.append({
            "field": ".".join(str(loc) for loc in error["loc"]),
            "message": error["msg"],
            "type": error["type"],
        })

    response = {
        "error": "Validation Error",
        "message": "Request validation failed",
        "statusCode": status.HTTP_400_BAD_REQUEST,
        "validation": errors,
    }

    if settings.is_development:
        response["detail"] = str(exc)

    return JSONResponse(
        status_code=status.HTTP_400_BAD_REQUEST,
        content=response,
    )


async def sqlalchemy_exception_handler(request: Request, exc: SQLAlchemyError):
    """
    Handle SQLAlchemy database errors.

    Returns 500 with generic message (hides details in production).
    """
    response = {
        "error": "Database Error",
        "message": "A database error occurred",
        "statusCode": status.HTTP_500_INTERNAL_SERVER_ERROR,
    }

    if settings.is_development:
        response["detail"] = str(exc)
        response["stack"] = exc.__class__.__name__

    return JSONResponse(
        status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
        content=response,
    )


async def generic_exception_handler(request: Request, exc: Exception):
    """
    Handle generic exceptions.

    Returns 500 with error details in development.
    """
    response = {
        "error": exc.__class__.__name__,
        "message": str(exc) or "An unexpected error occurred",
        "statusCode": status.HTTP_500_INTERNAL_SERVER_ERROR,
    }

    if settings.is_development:
        import traceback
        response["stack"] = traceback.format_exc()

    return JSONResponse(
        status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
        content=response,
    )

File: fastify-apps/checklist-template/src/plugins/errorHandler.ts
import { FastifyError, FastifyReply, FastifyRequest } from 'fastify';
import { ZodError } from 'zod';

import { ValidationError } from '../zod-schema-contract/common/index.js';

/**
 * Custom error classes for different error types
 */
export class NotFoundError extends Error {
  statusCode = 404;
  code = 'NOT_FOUND';

  constructor(message = 'Resource not found') {
    super(message);
    this.name = 'NotFoundError';
  }
}

export class ConflictError extends Error {
  statusCode = 409;
  code = 'CONFLICT';

  constructor(message = 'Resource conflict') {
    super(message);
    this.name = 'ConflictError';
  }
}

export class UnauthorizedError extends Error {
  statusCode = 401;
  code = 'UNAUTHORIZED';

  constructor(message = 'Unauthorized') {
    super(message);
    this.name = 'UnauthorizedError';
  }
}

export class ForbiddenError extends Error {
  statusCode = 403;
  code = 'FORBIDDEN';

  constructor(message = 'Forbidden') {
    super(message);
    this.name = 'ForbiddenError';
  }
}

/**
 * Global error handler for Fastify
 * Handles different error types and returns appropriate responses
 */
export function errorHandler(
  error: FastifyError | Error,
  request: FastifyRequest,
  reply: FastifyReply
): void {
  const isDevelopment = process.env.NODE_ENV === 'development';

  // Log the error with full details
  request.log.error({
    err: error,
    url: request.url,
    method: request.method,
    body: request.body,
    query: request.query,
    params: request.params,
  });

  // Handle Zod validation errors
  if (error instanceof ZodError) {
    void reply.status(400).send({
      success: false,
      error: {
        code: 'VALIDATION_ERROR',
        message: 'Validation failed',
        details: isDevelopment
          ? error.errors.map((err) => ({
              path: err.path.join('.'),
              message: err.message,
              code: err.code,
            }))
          : undefined,
      },
    });
    return;
  }

  // Handle custom ValidationError
  if (error instanceof ValidationError) {
    void reply.status(400).send({
      success: false,
      error: {
        code: 'VALIDATION_ERROR',
        message: error.message,
        details: isDevelopment ? error.toJSON() : undefined,
      },
    });
    return;
  }

  // Handle NotFoundError
  if (error instanceof NotFoundError) {
    void reply.status(404).send({
      success: false,
      error: {
        code: error.code,
        message: error.message,
      },
    });
    return;
  }

  // Handle ConflictError
  if (error instanceof ConflictError) {
    void reply.status(409).send({
      success: false,
      error: {
        code: error.code,
        message: error.message,
      },
    });
    return;
  }

  // Handle UnauthorizedError
  if (error instanceof UnauthorizedError) {
    void reply.status(401).send({
      success: false,
      error: {
        code: error.code,
        message: error.message,
      },
    });
    return;
  }

  // Handle ForbiddenError
  if (error instanceof ForbiddenError) {
    void reply.status(403).send({
      success: false,
      error: {
        code: error.code,
        message: error.message,
      },
    });
    return;
  }

  // Handle Fastify errors
  if ('statusCode' in error) {
    void reply.status(error.statusCode || 500).send({
      success: false,
      error: {
        code: error.code || 'INTERNAL_SERVER_ERROR',
        message: error.message,
        details: isDevelopment ? { stack: error.stack } : undefined,
      },
    });
    return;
  }

  // Handle generic errors
  void reply.status(500).send({
    success: false,
    error: {
      code: 'INTERNAL_SERVER_ERROR',
      message: isDevelopment ? error.message : 'An unexpected error occurred',
      details: isDevelopment ? { stack: error.stack } : undefined,
    },
  });
}
